fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexPrefetchTest.java

```diff
@@ -0,0 +1,483 @@
+
+        FDBRecordVersion version = storedRecord.getVersion();
+        assertThat(version.getLocalVersion(), equalTo(localVersion));
+    }
+
+    @Nonnull
+    private final RecordMetaDataHook simpleMetadataHook = metaDataBuilder -> {
+        // UseSplitRecords can be set to different values to impact the way the store is opened
+        metaDataBuilder.setSplitLongRecords(isUseSplitRecords());
+    };
+
+    @Nonnull
+    private RecordQueryPlan plan(final RecordQuery query, final RecordQueryPlannerConfiguration.IndexPrefetchUse useIndexPrefetch) {
+        planner.setConfiguration(planner.getConfiguration()
+                .asBuilder()
+                .setUseIndexPrefetch(useIndexPrefetch)
+                .build());
+        return planner.plan(query);
+    }
+
+    private byte[] executeAndVerifyData(RecordQueryPlan plan, int expectedRecords, BiConsumer<FDBQueriedRecord<Message>, Integer> recordVerifier) throws Exception {
+        return executeAndVerifyData(plan, null, ExecuteProperties.SERIAL_EXECUTE, expectedRecords, recordVerifier);
+    }
+
+    private byte[] executeAndVerifyData(RecordQueryPlan plan, byte[] continuation, ExecuteProperties executeProperties, int expectedRecords, BiConsumer<FDBQueriedRecord<Message>, Integer> recordVerifier) throws Exception {
+        int count = 0;
+        byte[] lastContinuation;
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, simpleMetadataHook);
+            try (RecordCursorIterator<FDBQueriedRecord<Message>> cursor = recordStore.executeQuery(plan, continuation, executeProperties).asIterator()) {
+                while (cursor.hasNext()) {
+                    FDBQueriedRecord<Message> record = cursor.next();
+                    recordVerifier.accept(record, count);
+                    count++;
+                }
+                lastContinuation = cursor.getContinuation();
+            }
+        }
+        assertThat(count, equalTo(expectedRecords));
+        return lastContinuation;
+    }
+
+    private List<FDBIndexedRawRecord> buildIndexedRawRecordWithOrphanIndexEntry() throws Exception {
+        long i = 9;
+        Index index = new Index("MyIndex", "a");
+        List<MappedKeyValue> mappedKeyValues = buildMappedResultWithOrphanIndexEntry();
+        List<FDBIndexedRawRecord> result = new ArrayList<>();
+        for (MappedKeyValue mappedKeyValue : mappedKeyValues) {
+            Tuple key = Tuple.from(991L, i--);
+            Tuple value = Tuple.from();
+            result.add(new FDBIndexedRawRecord(new IndexEntry(index, key, value), mappedKeyValue));
+        }
+        return result;
+    }
+
+    private List<MappedKeyValue> buildMappedResultWithOrphanIndexEntry() throws Exception {
+        List<MappedKeyValue> result = new ArrayList<>();
+        // Fully populated record
+        result.add(buildMappedKeyValue(new byte[] {21, 8, 20, 21, 17, 21, 2, 9}, new byte[0],
```

Comment:
Agree. This test was written to simulate an orphan index entry and the bytes were taken from a real record in order to simulate a realistic scenario. I intend to replace that test with a full end-to-end test that would create an orphan entry by disabling an index and then deleting a record. This test would be deleted at that point. Would it be OK to leave like that until then (not sure if we should delay this PR though)

Reply:
I'm fine with this going in as is, as long as the subspace prefix here is "reliable" (that is, it's not taken from, say, a directory layer that might differ on different machines, so the tests become flaky).

Reply:
Replaced the tests by an end-to-end test that creates a "real" orphan entry (using `uncheckedOpen`). This would also be useful for FDB testing as it verifies the behavior of the actual `getMappedRange` in this scenario.